### PR TITLE
feat(client): optional account recovery

### PR DESCRIPTION
**Pull request by fforbeck**
Issue https://github.com/storacha-network/project-tracking/issues/124

# Current Workflow (CLI)

### `w3cli` Project
The [`space.create(name, options)`](https://github.com/storacha-network/w3cli/blob/3f59da7b096a63f14def9946682160fefacd0702/space.js#L24) function in `space.js`  allows the user to pass the following options:
- **recovery**: `false` (default)
- **caution**: `false` (default)
- **customer**: `Email | false`
- **account**: `string | false`

If the `account` attribute is provided, the following steps are triggered:
1. `space.createdRecovery` is called.
2. Then, `client.capabilities.access.delegate` is executed.

This process occurs during **space creation**, not during provisioning.  
Provisioning only happens when the `customer` attribute is provided in the `options` argument.

---

# Issue (JS Client)

### `w3up` Project
The [`client.createSpace(name)`](https://github.com/storacha-network/w3up/blob/fb8b8677c4c633cdf8c259db55357a1794eed3ab/packages/w3up-client/src/client.js#L239) function in the `w3up` project currently does **not** accept any options, which means the recovery account must be created manually **after** space creation and provisioning. This introduces a risk of forgetting the manual step and potentially losing access to the space.

---

# Expected Outcome

We propose a small modification to the `client.createSpace(name)` function to support optional parameters, allowing clients to pass an `account` directly, automating the recovery setup.

### Solution:
- Clients can now call `client.createSpace(name, { account })` to include the `account` in the creation process.
- This ensures that the recovery account is created immediately after the space is created, provisioned, and saved, eliminating the need for a manual step.

### Benefits:
- **Backward Compatibility**: The change does not break existing client implementations since the `account` parameter is optional, keeping calls like `client.createSpace(name)` intact.
- **Simplified Workflow**: By simply passing the `account` when calling `createSpace`, the recovery account setup is handled automatically during space creation.
- **Flexibility**: If needed in the future, the `account` attribute can be made required to enforce stricter recovery account management.

### CODE REVIEW

File: packages/w3up-client/src/client.js [hunk 1]
```diff
@@ -233,12 +235,39 @@ export class Client extends Base {
...
+   * 
+   * @typedef {object} CreateOptions
+   * @property {string|false} [account]
+   * 
    * @param {string} name
+   * @param {CreateOptions} options
    */
-  async createSpace(name) {
-    return await this._agent.createSpace(name)
+  async createSpace(name, options = {}) {
```
Comment (hannahhoward) [inline packages/w3up-client/src/client.js:246]: at this API level, I think this should take an actual account object, rather than an email address. 

the code to obtain the account from the email can be left as an exercise above this code level.

also, the default flow for setting up the client often contains and instruction to login:

```
const myAccount = await client.login('zaphod@beeblebrox.galaxy')
```

So often the JS user already has the account.

**Review by hannahhoward — CHANGES_REQUESTED** [review]
This approach is fine, except I have one nit around using an account object rather than an email.

Also we should submit a PR to github.com/storacha-network/docs to update the docs for the JS client.

File: packages/w3up-client/test/client.test.js [hunk 2]
```diff
@@ -259,7 +259,33 @@ export const testClient = {
...
+      assert.ok(space)
+
+      const proof = alice.agent.proofs()
+      .find(p => p.issuer.did() === alice.did())
+      if (!proof) {
+        throw new Error('Recovery Proof not found')
+      }
+      
+      assert.ok(proof)
+      assert.equal(proof.audience.did(), alice.did())
```
Comment (fforbeck) [inline packages/w3up-client/test/client.test.js:282]: Thanks for the feedback @hannahhoward. I've applied the suggested changes. What is the best way to check if that account is a recovery account?

**Review by hannahhoward — APPROVED** [review]
Yea, this is obviously not ideal to have all this logic in creating a space, on the otherhand, with the chain of dependencies there's no way to seperate it. And it will make for a much shorter setup tutorial, which will be quite nice. We should definitely update docs for this.

**Review by hannahhoward — APPROVED** [review]
LGTM

**Comment by alanshaw** [comment]
@fforbeck how does this affect the space creation steps in the [CLI](https://github.com/storacha/w3cli)?

File: packages/w3up-client/src/client.js [hunk 3]
```diff
@@ -233,12 +235,50 @@ export class Client extends Base {
...
+      // Delegate space access to the recovery
+      const result = await this.capability.access.delegate({
+        space: space.did(),
+        delegations: [recovery],
+      })
+
+      if (result.error) {
+        throw new Error(
+          `⚠️ Failed to authorize recovery account: ${result.error.name}:${result.error.message}`
+        )
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:276]: I would just use the message in the error message, and add `result.error` as the cause for context. The `name` is largely for use by code and I don't think it adds anything to this message.

```suggestion
        throw new Error(`failed to authorize recovery account: ${result.error.message}`, { cause: result.error })
```

File: packages/w3up-client/src/client.js [hunk 4]
```diff
@@ -233,12 +235,50 @@ export class Client extends Base {
...
+    const space = await this._agent.createSpace(name)
+
+    const account = options.account
+    if (account) {
+      // Provision the account with the space
+      const provisionResult = await account.provision(space.did())
+      if (provisionResult.error) {
+        throw new Error(
+          `⚠️ Failed to provision account: ${provisionResult.error.name}:${provisionResult.error.message}`
+        )
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:258]: I recommend against using emojis in error messages - let a UI add that.

```suggestion
        throw new Error(`failed to provision account: ${provisionResult.error.message}`, { cause: provisionResult.error })
```

File: packages/w3up-client/src/client.js [hunk 5]
```diff
@@ -233,12 +235,50 @@ export class Client extends Base {
...
+      }
+
+      // Save the space to authorize the client to use the space
+      await space.save()
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:262]: Not sure about this - if I don't pass an `account` the space is not saved to my store...should it always be saved?

I think it's ok to only save if an account is passed but it needs to be mentioned in the jsdoc.

If you switch to always saving the space perhaps add a `noSave` option to retain the ability for ephemeral spaces to be created?

**Review by alanshaw — CHANGES_REQUESTED** [review]
This is a great but please could you try to keep non essential changes to a minimum? There's a lot of code formatting change in here that makes it harder for me to review. Just submit a separate PR in the future for code formatting changes 🙏.

File: packages/w3up-client/src/client.js (hunk 5 above)
Comment (fforbeck) [reply packages/w3up-client/src/client.js:260]: I believe it makes sense to save only if you provide the `account`, mainly because if you provide the account we need to get the authorizations done via `.save` call to be able to delegate to the recovery. Then if you want to create an ephemeral space you don't need the account because there is no reason to create a recovery. I will update the `jsdoc` to take that into account 😃

**Comment by fforbeck** [comment]
> This is a great but please could you try to keep non essential changes to a minimum? There's a lot of code formatting change in here that makes it harder for me to review. Just submit a separate PR in the future for code formatting changes 🙏.

Yeah, my bad. I will revert these formatting changes to simplify this PR.

**Comment by fforbeck** [comment]
> @fforbeck how does this affect the space creation steps in the [CLI](https://github.com/storacha/w3cli)?

The CLI should remain unaffected by this change because the `account` parameter is optional. The existing logic in the CLI already supports space creation without requiring an `account`, and it includes the necessary recovery logic (e.g., [this line in the code](https://github.com/storacha/w3cli/blob/main/space.js#L74)).

In the future, we could create a separate PR to simplify the CLI code by leveraging the new version of the `createSpace` function. However, given the current implementation, there should be no impact on the CLI’s functionality.

File: .github/workflows/w3up-client.yml [hunk 6]
```diff
@@ -31,8 +31,6 @@ jobs:
         uses: actions/checkout@v3
       - name: Install
         uses: pnpm/action-setup@v4
-        with:
-          version: 9
```
Comment (fforbeck) [inline .github/workflows/w3up-client.yml:35]: I've removed the version because it was causing conflicts with the version defined in the root package.json. It is optional when you have it in package.json: https://github.com/pnpm/action-setup?tab=readme-ov-file#version

File: packages/w3up-client/README.md [hunk 7]
```diff
@@ -126,55 +126,37 @@ A [`Space`][docs-Space] acts as a namespace for your uploads, and what your Agen
...
 
 ```js
 // wait for payment plan to be selected
 while (true) {
   const res = await account.plan.get()
   if (res.ok) break
   console.log('Waiting for payment plan to be selected...')
-  await new Promise(resolve => setTimeout(resolve, 1000))
+  await new Promise((resolve) => setTimeout(resolve, 1000))
 }
```
Comment (alanshaw) [inline packages/w3up-client/README.md:138]: It would be rad if this was a method of the client? Would you be up for adding something like that quickly?

**Review by alanshaw — APPROVED** [review]
LGTM

File: packages/w3up-client/README.md (hunk 7 above)
Comment (fforbeck) [reply packages/w3up-client/README.md:138]: Good idea. I will add that.

File: packages/w3up-client/src/client.js [hunk 8]
```diff
@@ -98,13 +99,45 @@ export class Client extends Base {
...
+      }
+
+      if (Date.now() - startTime > timeout) {
+        throw new Error('Timeout: Payment plan selection took too long.')
+      }
+
+      console.log('Waiting for payment plan to be selected...')
+      await new Promise((resolve) => setTimeout(resolve, interval))
+    }
+  }
```
Comment (fforbeck) [inline packages/w3up-client/src/client.js:139]: what about this @alanshaw ?

File: packages/w3up-client/README.md [hunk 9]
```diff
@@ -129,13 +129,8 @@ const account = await client.login('zaphod@beeblebrox.galaxy')
...
 ```js
-// wait for payment plan to be selected
-while (true) {
-  const res = await account.plan.get()
-  if (res.ok) break
-  console.log('Waiting for payment plan to be selected...')
-  await new Promise((resolve) => setTimeout(resolve, 1000))
-}
+// Wait for a payment plan with a 1-second polling interval and 60-second timeout
+await client.waitForPaymentPlan(account, 1000, 60000)
```
Comment (alanshaw) [inline packages/w3up-client/README.md:133]: 🙏 can we land as a separate PR? Apologies - I should have asked...

I think it might be better for this to be an instance method of `AccountPlan` rather than the top level? WDYT?

```js
await account.plan.wait()
```

In the docs I think it's simplier to just use the defaults (i.e. don't specify the interval/timeout).

File: packages/w3up-client/src/client.js [hunk 10]
```diff
@@ -107,6 +107,37 @@ export class Client extends Base {
...
 
+  /**
+   * Waits for a payment plan to be selected for the given account.
+   * This method continuously checks the account's payment plan status
+   * at a specified interval until a valid plan is selected or the timeout is reached.
+   *
+   * @param {Account.Account} account - The account object to check the payment plan for.
+   * @param {number} [interval=1000] - The polling interval in milliseconds (default is 1000ms).
+   * @param {number} [timeout=180000] - The maximum time to wait in milliseconds before throwing a timeout error (default is 3 minutes).
+   * @returns {Promise<void>} - Resolves once a payment plan is selected within the timeout.
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:118]: I would return the plan information when it is available.

File: packages/w3up-client/src/client.js [hunk 11]
```diff
@@ -107,6 +107,37 @@ export class Client extends Base {
     return Account.list(this)
   }
 
+  /**
+   * Waits for a payment plan to be selected for the given account.
+   * This method continuously checks the account's payment plan status
+   * at a specified interval until a valid plan is selected or the timeout is reached.
+   *
+   * @param {Account.Account} account - The account object to check the payment plan for.
+   * @param {number} [interval=1000] - The polling interval in milliseconds (default is 1000ms).
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:116]: I'd pass an options object with these two properties.

File: packages/w3up-client/src/client.js [hunk 12]
```diff
@@ -107,6 +107,37 @@ export class Client extends Base {
...
+   * @param {number} [timeout=180000] - The maximum time to wait in milliseconds before throwing a timeout error (default is 3 minutes).
```
Comment (alanshaw) [inline packages/w3up-client/src/client.js:117]: Might be better to pass an `AbortSignal` instead so that if folks are using this in a UI they can have a cancel button. Another reason is that it could take longer than 3 minutes, considering you have to log into your email, click the link, add credit card details etc.

File: packages/w3up-client/test/client.test.js [hunk 13]
```diff
@@ -698,6 +698,48 @@ export const testClient = {
...
+          plan: {
+            get: async () => ({ ok: false }),
+          },
+        }
+        // @ts-expect-error
+        await assert.rejects(alice.waitForPaymentPlan(account, 100, 500), {
+          message: 'Timeout: Payment plan selection took too long.',
+        })
+      },
+  },
```
Comment (alanshaw) [inline packages/w3up-client/test/client.test.js:742]: ❤️

**Review by alanshaw — CHANGES_REQUESTED** [review]

**Review by alanshaw — APPROVED** [review]
